presto-cli/src/main/java/com/facebook/presto/cli/StatusPrinter.java

```diff
@@ -443,4 +453,104 @@ private static double percentage(double count, double total)
+            this.timestampNanos = timestampNanos;
+            this.processedBytes = processedBytes;
+            this.processedRows = processedRows;
+        }
+
+        public long getTimestampNanos()
+        {
+            return timestampNanos;
+        }
+
+        public long getProcessedBytes()
+        {
+            return processedBytes;
+        }
+
+        public long getProcessedRows()
+        {
+            return processedRows;
+        }
+    }
+
+    private static class ProgressStatsHistory
+    {
+        private final Map<String, LinkedList<ProgressStats>> stageProgressStatsHistory = new HashMap<>();
+        private final LinkedList<ProgressStats> statementProgressStatsHistory = new LinkedList<>();
+        private final Duration statsHistoryInterval;
+        private final long start;
+
+        public ProgressStatsHistory(long timestampNanos, Duration statsHistoryInterval)
+        {
+            this.start = timestampNanos;
+            this.statementProgressStatsHistory.add(new ProgressStats(start, 0, 0));
+            this.statsHistoryInterval = statsHistoryInterval;
+        }
+
+        public void addStatementProgressStats(ProgressStats progressStats)
+        {
+            statementProgressStatsHistory.add(progressStats);
+            removeOutdatedStats(statementProgressStatsHistory);
+        }
+
+        public String formatStatementInstantDataRate()
+        {
+            return getFormattedDataRate(statementProgressStatsHistory);
+        }
+
+        public String formatStatementInstantRowCountRate()
+        {
+            return getFormattedRowCountRate(statementProgressStatsHistory);
+        }
+
+        public void addStageProgressStats(String id, ProgressStats progressStats)
+        {
+            if (!stageProgressStatsHistory.containsKey(id)) {
+                stageProgressStatsHistory.put(id, new LinkedList<>());
+                stageProgressStatsHistory.get(id).add(new ProgressStats(start, 0, 0));
+            }
+            LinkedList<ProgressStats> stageHistory = stageProgressStatsHistory.get(id);
+            stageHistory.add(progressStats);
+            removeOutdatedStats(stageHistory);
+        }
+
+        public String formatStageInstantDataRate(String id)
+        {
+            return getFormattedDataRate(stageProgressStatsHistory.get(id));
+        }
+
+        public String formatStageInstantRowCountRate(String id)
+        {
+            return getFormattedRowCountRate(stageProgressStatsHistory.get(id));
+        }
+
+        private void removeOutdatedStats(LinkedList<ProgressStats> progressStatsHistory)
+        {
+            while (succinctNanos(progressStatsHistory.getLast().getTimestampNanos() - progressStatsHistory.getFirst().getTimestampNanos()).compareTo(statsHistoryInterval) > 0) {
+                progressStatsHistory.removeFirst();
+            }
+        }
+
+        private static String getFormattedDataRate(LinkedList<ProgressStats> progressStatsHistory)
```

Comment:
This can be a static method in the parent class, doesn't fit well in this class.